dash_bio/utils/gene_expression_reader.py

```diff
@@ -1,57 +1,151 @@
 
-        for c in columns:
-            if c not in all_cols:
-                continue
-            selected_cols.append(c)
+        return desc, subsets, all_rows, all_cols
 
-        selected_data = df.loc[selected_rows, selected_cols]
-        data = selected_data.values
+    return _get_selected_data(
+        df,
+        all_rows,
+        all_cols,
+        rows,
+        columns
+    )
 
+
+def read_tsv_file(
+        contents='',
+        filepath='',
+        rows=None,
+        columns=None,
+        description_row_prefix='#',
+        description_separator=':',
+        skiprows=0,
+        ignore_columns=None,
+        index_column='Gene Name',
+        return_filtered_data=False
+):
     desc = {}
-    info = pd.read_csv(filepath, sep='^', nrows=header_rows-1, header=None)[0]
-    for i in info:
-        tmp = i.strip('#').split(':', 1)
-        if len(tmp) < 2:
-            desc['Source'] = tmp[0]
+
+    if len(contents) > 0:
+        with tempfile.NamedTemporaryFile(
+                mode='w+', delete=False, suffix='.soft'
+        ) as tf:
+            tf.write(contents)
+            filepath = tf.name
+
+        for line in tf.lines:
+            if line[0] == description_row_prefix or len(line.strip()) == 0:
+                if line[0] == description_row_prefix:
+                    desc_line = line.strip(
+                        description_row_prefix
+                    ).strip().split(description_separator, 1)
+                    desc[desc_line[0]] = desc_line[1]
+                skiprows += 1
+            else:
+                break
+
+    else:
+        with open(filepath, 'r') as f:
+            for line in f.readlines():
+                if line[0] == description_row_prefix or len(line.strip()) == 0:
+                    if line[0] == description_row_prefix:
+                        desc_line = line.strip(
+                            description_row_prefix
+                        ).strip().split(description_separator, 1)
+                        desc[desc_line[0]] = desc_line[1]
+                        skiprows += 1
+                else:
+                    break
+
+    df = pd.read_csv(filepath, sep='\t', skiprows=skiprows)
+
+    df.set_index(index_column, inplace=True)
+
+    all_rows = list(df.index.values)
+    all_cols = list(df.columns.values)
+
+    for ignore_col in ignore_columns:
+        all_cols.remove(ignore_col)
+
+    if not return_filtered_data:
+        return desc, all_rows, all_cols
+
+    return _get_selected_data(
+        df,
+        all_rows,
+        all_cols,
+        rows,
+        columns
+    )
+
+
+def _get_selected_data(
+        dataframe,
+        all_rows=None,
+        all_cols=None,
+        rows=None,
+        columns=None
+):
+    if all_rows is None:
+        all_rows = []
+    if all_cols is None:
+        all_cols = []
+    if rows is None:
+        rows = []
+    if columns is None:
+        columns = []
+
+    selected_rows = []
```

Comment:
```suggestion
    selected_rows = list(set(all_rows).intersection(rows))
```
and discard the following loops... no?

Reply:
Good catch!

Reply:
Fixed in 59260c4